x-pack/platform/plugins/shared/ml/server/routes/trained_models.ts

```diff
@@ -74,17 +75,30 @@ export function trainedModelsRoutes(
         version: '1',
         validate: false,
       },
-      routeGuard.fullLicenseAPIGuard(async ({ client, mlClient, request, response }) => {
-        try {
-          const modelsClient = modelsProvider(client, mlClient, cloud, getEnabledFeatures());
-          const models = await modelsClient.getTrainedModelList();
-          return response.ok({
-            body: models,
-          });
-        } catch (e) {
-          return response.customError(wrapError(e));
+      routeGuard.fullLicenseAPIGuard(
+        async ({ client, mlClient, request, response, mlSavedObjectService }) => {
+          try {
+            const modelsClient = modelsProvider(client, mlClient, cloud, getEnabledFeatures());
+            const { trainedModelsSpaces } = checksFactory(client, mlSavedObjectService);
+
+            const [models, spaces] = await Promise.all([
+              modelsClient.getTrainedModelList(),
+              trainedModelsSpaces(),
+            ]);
+
+            const mergedModels = models.map((model) => ({
+              ...model,
+              spaces: spaces.trainedModels[model.model_id] ?? [],
+            }));
```

Comment:
Did you consider passing `mlSavedObjectService` to the `modelsProvider` and then resolving `trainedModelsSpaces` internally there? 
```suggestion
            const modelsClient = modelsProvider(client, mlClient, mlSavedObjectService, cloud, getEnabledFeatures(), );
            const models = await modelsClient.getTrainedModelList();
```

Reply:
Refactored to assign spaces inside `modelsProvider` in https://github.com/elastic/kibana/pull/213567/commits/b11f1fe6738bcd0e19460fcd86cfaea0fbda79f1